hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -802,29 +701,249 @@ update env ledger st ev = case (st, ev) of
+        { parameters = parameters
+        , pendingCommits = Set.fromList parties
+        , committed = mempty
+        , chainState
+        , headId
+        }
+  CommittedUTxO{committedUTxO, chainState, party} ->
+    case st of
+      Initial InitialState{parameters, pendingCommits, committed, headId} ->
+        Initial
+          InitialState
+            { parameters
+            , pendingCommits = remainingParties
+            , committed = newCommitted
+            , chainState
+            , headId
+            }
+       where
+        newCommitted = Map.insert party committedUTxO committed
+        remainingParties = Set.delete party pendingCommits
+      _otherState -> st
+  TransactionReceived{tx} ->
+    case st of
+      Open os@OpenState{coordinatedHeadState} ->
+        Open
+          os
+            { coordinatedHeadState =
+                -- Spec: Tall ← ̂Tall ∪ { (hash(tx), tx) }
+                coordinatedHeadState
+                  { allTxs = allTxs <> fromList [(txId tx, tx)]
+                  }
+            }
+       where
+        CoordinatedHeadState{allTxs} = coordinatedHeadState
+      _otherState -> st
+  TransactionAppliedToLocalUTxO{tx, utxo} ->
+    case st of
+      Open os@OpenState{coordinatedHeadState} ->
+        Open
+          os
+            { coordinatedHeadState =
+                coordinatedHeadState
+                  { localUTxO = utxo
+                  , localTxs = localTxs <> [tx]
+                  }
+            }
+       where
+        CoordinatedHeadState{localTxs} = coordinatedHeadState
+      _otherState -> st
+  SnapshotRequestDecided{snapshotNumber} ->
+    case st of
+      Open os@OpenState{coordinatedHeadState} ->
+        Open
+          os
+            { coordinatedHeadState =
+                coordinatedHeadState
+                  { seenSnapshot =
+                      RequestedSnapshot
+                        { lastSeen = seenSnapshotNumber seenSnapshot
+                        , requested = snapshotNumber
+                        }
+                  }
+            }
+       where
+        CoordinatedHeadState{seenSnapshot} = coordinatedHeadState
+      _otherState -> st
+  SnapshotRequested{snapshot, requestedTxIds} ->
+    case st of
+      Open os@OpenState{coordinatedHeadState, currentSlot} ->
+        Open
+          os
+            { coordinatedHeadState =
+                coordinatedHeadState
+                  { seenSnapshot = SeenSnapshot snapshot mempty
+                  , localTxs = localTxs'
+                  , localUTxO = localUTxO'
+                  , allTxs = foldr Map.delete allTxs requestedTxIds
+                  }
+            }
+       where
+        (localTxs', localUTxO') = pruneTransactions snapshotUtxo
+        Snapshot{utxo = snapshotUtxo} = snapshot
+
+        CoordinatedHeadState{allTxs, localTxs} = coordinatedHeadState
+        pruneTransactions utxo = do
+          foldr go ([], utxo) localTxs
+         where
+          go tx (txs, u) =
```

Comment:
This is the only place in the `aggregate` function we use the `ledger` and this is a smell to me we should get rid of this dependency here:
* Events represent "past" things that transform the state directly and do not require additional context,
* The ledger is a complex thing to build, it requires parameters which might not be trivial to retrieve and might even change over time within a head (we don't support this feature but it could be),
* Applying transactions is probably the least efficient part of the whole event interpretation,
* The main advantage of event sourcing is that it provides a stream of "raw" data for downstream consumers, possibly even with vastly different usage, 
* events should be self-contained enough to provide all the data needed to reconstruct the state.

For all these reasons, I would like to suggest we rewrite this part to have the `SnapshotRequested` event contain the _resolved_ transactions and changes in UTxO, something like:

```
   SnapshotRequested { snapshot :: Snapshot, localTxs :: [tx], localUTxO :: UTxOType tx }
```

Because the `localUTxO` could be large, alternatively we could have:

```
   SnapshotRequested { snapshot :: Snapshot, remainingTxs :: [tx], addedUTxO :: UTxOType tx, deletedUTxO :: UTxOType tx }
```

Reply:
We had it the other way and moved it here to keep the events minimal in what they contain. i.e. persist only the minimum of what is required to reconstruct the state.

> Events represent "past" things that transform the state directly

But having a `StateReplaced` event is undesirable still, no? So I thought we should be keeping events small and to the point.

> The ledger is a complex thing to build, it requires parameters which might not be trivial to retrieve and might even change over time within a head

Even more a reason to keep it in the aggregate IMO as those new / current parameters would be in the state if they can change

> The main advantage of event sourcing is that it provides a stream of "raw" data for downstream consumers, possibly even with vastly different usage,

Correct, and not all downstream users (aggregates, policies) have a notion of `localUTxO` or `localTxs`. So to them it would even be noise to keep these things in the event?

> events should be self-contained enough to provide all the data needed to reconstruct the state

Why would you think this is not the case anymore?

Reply:
I also prefer keeping the ledger out of `aggregate` function. Then `aggregate` really doesn't depend on anything special and only does some data shuffling between different records.

Reply:
We did it this way now again.